### PR TITLE
Use `edition = "2024"` in the compiler

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory.

This involves changing all of the compiler crates' edition to `2024` (except for `codegen_gcc` and `codegen_cranelift` and `pattern_analysis`), and massively simplifying opaques and their lifetimes. I think the overwhelming majority of functions look significantly better now.

There are some other changes sprinkled in too, such as how env var functions are now unsafe.

### CODE REVIEW

**Review by traviscross — APPROVED** [review]
This all looks right to me.

It's gratifying to see how much the Lifetime Capture Rules 2024 really clean things up, in particular by allowing lifetime elision to be used in many more places.

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs [hunk 1]
```diff
@@ -114,7 +114,7 @@ enum Scope<'a> {
         /// If this binder comes from a where clause, specify how it was created.
         /// This is used to diagnose inaccessible lifetimes in APIT:
         /// ```ignore (illustrative)
-        /// fn foo(x: impl for<'a> Trait<'a, Assoc = impl Copy + 'a>) {}
+        /// fn foo(x: impl for<'a> Trait<'a, Assoc = impl Copy >) {}
```
[the remark this answers is not in the document]
Comment (traviscross) [reply compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:117]: Could also mark the example as `edition2024,ignore` to make clear that the lifetime is captured.  Otherwise, or to be explicit, the `+ use<'a>` you suggest makes more sense to me.

**Comment by compiler-errors** [comment]
I think this is ready.

**Comment by compiler-errors** [comment]
r? compiler

File: compiler/rustc_pattern_analysis/Cargo.toml [hunk 2]
```diff
@@ -1,7 +1,9 @@
+cargo-features = ["edition2024"]
+
 [package]
 name = "rustc_pattern_analysis"
 version = "0.0.0"
-edition = "2021"
+edition = "2024"
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/Cargo.toml:6]: I guess it may be worthwhile to revert the changes to pattern analysis if R-A needs it to build.

**Review by lcnr — COMMENTED** [review]
can we remove `Captures` now that we have explicit `use` syntax?

I thought `tidy` requires safety comments for unsafe blocks? 🤔 I feel somewhat :/ about wrapping `set_var` in an `unsafe` block without explaining why that's correct. I guess it's always `SAFETY: rustc is single threaded up until this point`? I am not totally confident that this holds for all `unsafe` blocks, so alternatively, please open an issue tracking the fact that we may unsoundly use `[set|remove]_env` in the compiler

**Comment by lcnr** [comment]
apart from that r=me, cc @rust-lang/compiler @rust-lang/compiler-contributors

**Comment by compiler-errors** [comment]
> can we remove Captures now that we have explicit use syntax?

I'll have to wait until `use<...>` is stable for that to be doable. I've hestitated to touch anything having to do with r-a for stability reasons. I can open a draft PR on top of this converting the last crate to, and deleting `Captures` from the pattern analysis crate outright.

>  I am not totally confident that this holds for all unsafe blocks, so alternatively, please open an issue tracking the fact that we may unsoundly use [set|get]_env in the compiler

That's a good point. Yeah, I'll open a tracking issue for that (and enforcing `// SAFETY` in the compiler in general). and see if anyone would like to chime in on the safety. I'm not totally qualified, but also this is no more unsafe than it was on edition 2021, just more explicit.

**Comment by saethlin** [comment]
> I thought `tidy` requires safety comments for unsafe blocks? 🤔 

We have a handful of unsafe blocks without comments, such as https://github.com/rust-lang/rust/blob/a5cf8bbd4e1c8edeae08778c85c6f806dd00e853/compiler/rustc_index/src/slice.rs#L137-L138
https://github.com/rust-lang/rust/blob/a5cf8bbd4e1c8edeae08778c85c6f806dd00e853/compiler/rustc_data_structures/src/tagged_ptr/copy.rs#L153-L155

Considering the strange way we handle formatting, I wouldn't speculate about what tidy is doing here.

File: compiler/rustc_pattern_analysis/Cargo.toml (hunk 2 above)
Comment (lqd) [reply compiler/rustc_pattern_analysis/Cargo.toml]: I can try when I’m back from holidays but do you know if the rustc-crates-on-stable rmake test failed as we expect with the pattern analysis changes? I wonder if using some of cargo’s unstable would make it fail. If not, I’d like to fix it.

**Comment by RalfJung** [comment]
> but also this is no more unsafe than it was on edition 2021, just more explicit.

Well the entire point of this is to give everyone a chance to audit their use of `set_env`. If not even rustc bothers with thta, we can hardly expect any better from our users...

File: compiler/rustc_interface/src/passes.rs [hunk 3]
```diff
@@ -182,13 +182,15 @@ fn configure_and_expand(
...
             }
-            env::set_var(
-                "PATH",
-                &env::join_paths(
-                    new_path.iter().filter(|p| env::join_paths(iter::once(p)).is_ok()),
-                )
-                .unwrap(),
-            );
+            unsafe {
+                env::set_var(
```
Comment (RalfJung) [inline compiler/rustc_interface/src/passes.rs:186]: This one is Windows-only where the underlying operation is *not* racy -- so that can probably be justified with a comment immediately.

File: compiler/rustc_llvm/build.rs [hunk 4]
```diff
@@ -51,9 +51,13 @@ fn detect_llvm_link() -> (&'static str, &'static str) {
 fn restore_library_path() {
     let key = tracked_env_var_os("REAL_LIBRARY_PATH_VAR").expect("REAL_LIBRARY_PATH_VAR");
     if let Some(env) = tracked_env_var_os("REAL_LIBRARY_PATH") {
-        env::set_var(&key, env);
+        unsafe {
```
Comment (RalfJung) [inline compiler/rustc_llvm/build.rs:54]: This one (and the others that are not immediately obvious) should get a FIXME, like the warning suggests.

File: compiler/rustc_feature/src/tests.rs [hunk 5]
```diff
@@ -3,7 +3,9 @@ use super::UnstableFeatures;
 #[test]
 fn rustc_bootstrap_parsing() {
     let is_bootstrap = |env, krate| {
-        std::env::set_var("RUSTC_BOOTSTRAP", env);
+        unsafe {
+            std::env::set_var("RUSTC_BOOTSTRAP", env);
```
Comment (bjorn3) [inline compiler/rustc_feature/src/tests.rs:7]: This would affect other concurrent tests, right?

File: compiler/rustc_interface/src/passes.rs (hunk 3 above)
Comment (WaffleLapkin) [reply compiler/rustc_interface/src/passes.rs:186]: Makes me think if we should have `std::os::windows::set_var` which is safe 🤔

File: compiler/rustc_hir_analysis/src/variance/mod.rs [hunk 6]
```diff
@@ -77,11 +77,8 @@ fn variances_of(tcx: TyCtxt<'_>, item_def_id: LocalDefId) -> &[ty::Variance] {
 fn variance_of_opaque(tcx: TyCtxt<'_>, item_def_id: LocalDefId) -> &[ty::Variance] {
     let generics = tcx.generics_of(item_def_id);
 
-    // Opaque types may only use regions that are bound. So for
-    // ```rust
-    // type Foo<'a, 'b, 'c> = impl Trait<'a> + 'b;
-    // ```
-    // we may not use `'c` in the hidden type.
+    /// Collect the lifetimes that are mentioned in the opaque that come from the parent,
+    /// and mark them as invariant.
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/variance/mod.rs:81]: The comment was wrong.

**Comment by compiler-errors** [comment]
I've opened https://github.com/rust-lang/rust/issues/130672 to track those couple sketchy usages of `set_var`. I think this is otherwise good to go.

@bors r=lcnr

**Comment by lqd** [comment]
> do you know if the `rustc-crates-on-stable` rmake test failed as we expect with the pattern analysis changes? I wonder if using some of cargo’s unstable features would make it fail.

The answer is no, unfortunately. 

It's also unclear to me whether we can ask cargo to only check the unstable features in the crates _we're building_ rather than all the crates in the workspace. I've tried and couldn't do it: the presence of the unstable edition 2024 feature anywhere + cargo's `-Zallow-features=` failed validation even on the edition 2021 crates.

TL;DR: to my knowledge, we can't easily enforce on CI to not use cargo unstable features (including edition 2024) in the following crates that need to build on stable:
- rustc_type_ir
- rustc_next_trait_solver
- rustc_pattern_analysis
- rustc_lexer

**Comment by compiler-errors** [comment]
Hm... I probably should hold back rustc_lexer from edition 2024 as well.

@bors r-

**Comment by lqd** [comment]
> I probably should hold back rustc_lexer from edition 2024

as well as these 2 crates I've just learned r-a also uses (via #131997)
- `rustc_abi`
- `rustc_parse_format`

**Comment by Nadrieril** [comment]
r-a also uses `rustc_pattern_analysis` and `rustc_index` I believe.

**Comment by WaffleLapkin** [comment]
@compiler-errors did this land in some other form/why is this closed?

**Comment by lqd** [comment]
IIUC https://github.com/rust-lang/rust/pull/137333 and https://github.com/rust-lang/rust/pull/137334 replaced it.
